keras_nlp/samplers/contrastive_sampler.py

```diff
@@ -0,0 +1,232 @@
+            flat_shape = [batch_size * self.k] + x.shape.as_list()[1:]
+            return tf.reshape(x, shape=flat_shape)
+
+        def flatten_beams(x):
+            """Combine the beam dim and batch dim."""
+            flat_shape = [batch_size * self.k] + x.shape.as_list()[2:]
+            return tf.reshape(x, shape=flat_shape)
+
+        def unflatten_beams(x):
+            """Separate the beam dim and batch dim."""
+            unflat_shape = [batch_size, self.k] + x.shape.as_list()[1:]
+            return tf.reshape(x, shape=unflat_shape)
+
+        mask = tf.zeros_like(prompt, dtype=tf.bool) if mask is None else mask
+        # Compute initial logits.
+        logits, _, cache = next(prompt, cache, index)
+        # `tf.while_loop` will not accept `None` as a value for `loop_vars`.
+        cache = () if cache is None else cache
+
+        def cond(prompt, cache, index, logits, hidden_states):
+            if end_token_id is None:
+                return True
+            # Stop if all sequences have produced a *new* end_token_id.
+            end_tokens = (prompt == end_token_id) & (~mask)
+            prompt_done = tf.reduce_any(end_tokens, axis=-1)
+            return not tf.reduce_all(prompt_done)
+
+        def body(prompt, cache, index, logits, hidden_states):
+            # Compute the softmax distribution for the next token.
+            probabilities = keras.activations.softmax(logits)
+
+            # Replicate for `self.k` times to find the best token in top-k
+            # candidates.
+            prompt_beams = create_beams(prompt)
+            mask_beams = create_beams(mask)
+            hidden_states_beams = create_beams(hidden_states)
+            cache_beams = tf.nest.map_structure(create_beams, cache)
+
+            # Get top-k candidate tokens and their probabilities.
+            top_k_probabilities, top_k_indices = tf.math.top_k(
+                probabilities, k=self.k, sorted=False
+            )
+            next_token_probabilities = flatten_beams(top_k_probabilities)
+            next_token = flatten_beams(top_k_indices)
+            next_token = tf.cast(next_token, prompt.dtype)
+            next_token = tf.where(
+                mask_beams[:, index], prompt_beams[:, index], next_token
+            )
+
+            # Update the prompt with the next token.
+            next_token = next_token[:, tf.newaxis]
+            prompt_beams = dynamic_update_slice(
+                prompt_beams, next_token, [0, index]
+            )
+
+            # Compute the logits and hidden states for top-k candidate tokens.
+            next_logits, next_hidden_states_beams, cache_beams = next(
+                prompt_beams, cache_beams, index + 1
+            )
+
+            # Compute the max similarity score for top-k candidate tokens
+            # against previous tokens.
+            last_token_state = next_hidden_states_beams
+            previous_states = hidden_states_beams[:, :index, :]
```

Comment:
Have we benchmarked this against an approach where we do not use a dynamically lengthed tensor, and instead mask out (by setting to a really low number) the similarity scores for indices above our current?

I am always afraid to do things dynamically length with XLA, because as I understand it will be choosing some (maybe way to large) heuristic length in this case.

Reply:
Creating a mask tensor inside XLA graph is a bit tricky, I ended up doing slice at the `similarity_score`, which techinically should be faster?